cleanrl/muesli_atari_envpool_async_jax_scan_impalanet_machado.py

```diff
@@ -0,0 +1,1393 @@
+    if callable(learning_rate):
+        return optax.scale_by_schedule(lambda count: m * learning_rate(count))
+    return optax.scale(m * learning_rate)
+
+
+def clipped_adamw(
+    learning_rate: optax.ScalarOrSchedule,
+    b1: float = 0.9,
+    b2: float = 0.999,
+    eps: float = 1e-8,
+    eps_root: float = 0.0,
+    mu_dtype: tp.Any | None = None,
+    weight_decay: float = 1e-4,
+    mask: tp.Optional[tp.Union[tp.Any, tp.Callable[[optax.Params], tp.Any]]] = None,
+    max_parameter_update: float = 1.0,
+) -> optax.GradientTransformation:
+    r"""The AdamW optimizer, except we clip the gradient before scaling by the learning rate.
+
+    See Hessel et al. 2022, Muesli paper, Muesli Supplement F.1: Optimizers for details.
+
+    Args:
+        learning_rate: A fixed global scaling factor.
+        b1: Exponential decay rate to track the first moment of past gradients.
+        b2: Exponential decay rate to track the second moment of past gradients.
+        eps: A small constant applied to denominator outside of the square root
+          (as in the Adam paper) to avoid dividing by zero when rescaling.
+        eps_root: A small constant applied to denominator inside the square root (as
+          in RMSProp), to avoid dividing by zero when rescaling. This is needed for
+          example when computing (meta-)gradients through Adam.
+        mu_dtype: Optional `dtype` to be used for the first order accumulator; if
+          `None` then the `dtype` is inferred from `params` and `updates`.
+        weight_decay: Strength of the weight decay regularization. Note that this
+          weight decay is multiplied with the learning rate. This is consistent
+          with other frameworks such as PyTorch, but different from
+          (Loshchilov et al, 2019) where the weight decay is only multiplied with
+          the "schedule multiplier", but not the base learning rate.
+        mask: A tree with same structure as (or a prefix of) the params PyTree,
+          or a Callable that returns such a pytree given the params/updates.
+          The leaves should be booleans, `True` for leaves/subtrees you want to
+          apply the weight decay to, and `False` for those you want to skip. Note
+          that the Adam gradient transformations are applied to all parameters.
+        max_parameter_update: The amount to clip the parameter update by.
+    Returns:
+        The corresponding `GradientTransformation`.
+    """
+    return optax.chain(
+        optax.scale_by_adam(b1=b1, b2=b2, eps=eps, eps_root=eps_root, mu_dtype=mu_dtype),
+        optax.add_decayed_weights(weight_decay, mask),
+        optax.clip(max_parameter_update),
+        scale_by_learning_rate(learning_rate),
+    )
+
+
+def update_target_network(state: TrainState, target_state: TrainState, alpha_target: float) -> TrainState:
+    new_target_params = jax.tree_multimap(
+        lambda p, tp: p * alpha_target + tp * (1 - alpha_target), state.params, target_state.params
```

Comment:
It may be easier to use `target_params=optax.incremental_update(qf1_state.params, qf1_state.target_params, args.tau)`

Reply:
Fixed.

Reply:
(Although the source for `optax.incremental_update` is basically that. Better not to re-invent the wheel though.)